apiserver/src/webhook/mod.rs

```diff
@@ -0,0 +1,316 @@
+    let desired_version = request.desired_api_version.clone();
+
+    match convert_objects(desired_version, request.objects.clone()) {
+        Ok(new_objects) => {
+            let response = Response {
+                uid: request.uid.clone(),
+                result: ConvertResult::default(),
+                converted_objects: Some(new_objects),
+            };
+            ConversionResponse {
+                kind: req.kind.clone(),
+                api_version: req.api_version.clone(),
+                response: response,
+            }
+        }
+        Err(e) => {
+            let fail_result = ConvertResult::create_fail_result(e.to_string());
+            let response = Response {
+                uid: request.uid.clone(),
+                result: fail_result,
+                converted_objects: None,
+            };
+            ConversionResponse {
+                kind: req.kind.clone(),
+                api_version: req.api_version.clone(),
+                response: response,
+            }
+        }
+    }
+}
+
+#[instrument(err)]
+fn convert_objects(
+    desired_version: String,
+    objects: Vec<serde_json::Value>,
+) -> Result<Vec<serde_json::Value>> {
+    let mut new_objects = Vec::new();
+    for old_object in objects.into_iter() {
+        let old_brs_object = BRSObject { object: old_object };
+        let new_brs_object = old_brs_object.chained_convert_object(desired_version.clone())?;
+        new_objects.push(new_brs_object.object);
+    }
+    Ok(new_objects)
+}
+
+struct BRSObject {
+    pub object: serde_json::Value,
+}
+
+impl BRSObject {
+    fn get_version(&self) -> Result<String> {
+        Ok(serde_json::from_value(self.object["apiVersion"].clone())
+            .context(SourceVersionNotExistInRequest)?)
+    }
+
+    fn to_v2(source_obj: BRSObject) -> Result<BRSObject> {
+        Self::try_from(BottlerocketShadow_v2::from(
+            BottleRocketShadow_v1::try_from(source_obj)?,
+        ))
+    }
+
+    fn convert_to_next_version(self) -> Result<Self> {
+        let version = self.get_version()?;
+        match version.as_str() {
+            "brupop.bottlerocket.aws/v1" => BRSObject::to_v2(self),
+            _ => InvalidVersionError { version }.fail(),
+        }
+    }
+
+    #[instrument(skip(self), err)]
+    fn chained_convert_object(self, desired_version: String) -> Result<Self> {
+        let mut version = self.get_version()?;
+        let mut source_object = self;
+
+        while version != desired_version {
+            match source_object.convert_to_next_version() {
+                Ok(val) => source_object = val,
+                Err(_) => {
+                    return ChainedConvertError {
+                        src_version: version,
+                        dst_version: desired_version,
+                    }
+                    .fail()
+                }
+            }
+            version = source_object.get_version()?;
+        }
+
+        Ok(source_object)
```

Comment:
```suggestion
        let mut source_object = self;
        while version.get_version()? != desired_version {
            // ChainedConverError should just add more context to the error returned by "convert_to_next_version"
            source_object = source_object.convert_to_next_version().context(ChainedConvertErrorSnafu)?;
        }

        Ok(source_object)

```

Reply:
I don't think I am apply what you suggested here. 

First of all, `convert_to_next_version` returns error `InvalidVersionError` which indicate that the next version doesn't exist in the map, but it didn't have any context on the target version. So I need `ChainedConvertError` to give more information on during the chained convert, which version it is in and what is the target version.

I am not using context here because both `InvalidVersionError` and `ChainedConvertError` are defined in `WebhookConvertError` and I really couldn't find a way to put `InvalidVersionError` as source for `ChainedConvertError`. One way could be I put `InvalidVersionError` to another error type, but I don't know whether it would be worthy to do that.